text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+```rust
+bitfield! {
+    struct MipsInstruction {
+        opcode: u<6>,
+        rs: u<5>,
+        rt: u<5>,
+        rd: u<5>,
+        shift: u<5>,
+        function: u<6>,
+    }
+}
+```
+
+Which would roughly be equivalent to:
+
+```rust
+struct MipsInstruction { /* ... */ }
+impl MipsInstruction {
+    fn pack(opcode: u<6>, rs: u<5>, rt: u<5>, rd: u<5>, shift: u<5>, function: u<6>) -> MipsInstruction { /* ... */ }
+    fn opcode(&self) -> u<6> { /* ... */ }
+    fn rs(&self) -> u<5> { /* ... */ }
+    fn rt(&self) -> u<5> { /* ... */ }
+    fn rd(&self) -> u<5> { /* ... */ }
+    fn shift(&self) -> u<5> { /* ... */ }
+    fn function(&self) -> u<6> { /* ... */ }
+}
+```
+
+Having the ability to explicitly state in your API how many bits a field takes up, and to be able to statically ensure that someone is providing exactly that many bits, is a pretty nice thing to have. Without generic integers, we can't make those kinds of API guarantees.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Up until now, we've only seen *specific* integer types, like `u8` or `i32`. What if we want to write a trait that works with all of them?
+
+Let's go with a simple example. Let's say that we want to code our own `Default` trait which returns one instead of zero. We'll define it like this:
+
+```rust
+pub trait MyDefault: Sized {
+    fn my_default() -> Self;
+}
+```
+
+Right now, with what we know, we have to copy-paste our impl for every integer type. For example, an impl for `u8` would be:
+
+```rust
+impl MyDefault for u8 {
+    fn my_default() -> u8 { 1 }
+}
+```
+
+Except, we'd have to replicate this for every single integer type. If we're clever, we could use a macro:
+
+```rust
+macro_rules! impl_my_default {
+    ($($i: ident),*) => {$(
+        impl MyDefault for $i {
+            fn my_default() -> $i { 1 }
+        }
+    )*}
+}
+impl_my_default!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+```
+
+But really, that's just making the compiler do the copy-pasting for us. Instead, we're going to use the special types `u<N>` and `i<N>` to generalise the code for us. The end result looks like:
+
+```rust
+impl<const N: usize> MyDefault for u<N> {
+    fn my_default() -> u<N> { 1 }
+}
+impl<const N: usize> MyDefault for i<N> {
+    fn my_default() -> i<N> { 1 }
+}
+impl MyDefault for usize {
+    fn my_default() -> usize { 1 }
+}
+impl MyDefault for isize {
+    fn my_default() -> isize { 1 }
+}
+```
+
+That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.
```

Comment:
```suggestion
That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 32-bit system, `usize` is not the same as `u64`.
```

Or any other combination, since on a 64-bit system usize is indeed the same as u64 :smile:

Reply:
No, this is the correct wording.

Even though they're the same layout in memory, they're different types intentionally so that you don't need to distinguish this depending on the actual layout for the specific system.